### PR TITLE
Feature : Travel activities

### PR DESCRIPTION
**Pull request by yassine04e**
feat: Adding the Activity class, the repositories, model view. Adding also the screens to add an activity to a travel, and the listing of all activities

### CODE REVIEW

File: app/src/main/java/com/github/se/travelpouch/model/activity/Activity.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package com.github.se.travelpouch.model.activity
+
+import com.github.se.travelpouch.model.Location
+import com.google.firebase.Timestamp
+
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/model/activity/Activity.kt:5]: You can add documentation to explain what an activity is

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+     * @param (() -> Unit) : the function to apply when the authentication goes without a trouble
+     */
+    override fun init(onSuccess: () -> Unit) {
+        Firebase.auth.addAuthStateListener {
+            if (it.currentUser != null) {
+                onSuccess()
+            }
+        }
+    }
+
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt:37]: You can add comments for these function. It will clarify their purpose and usage.

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    override fun updateActivity(
+        activity: Activity,
+        onSuccess: () -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+        Log.d("TravelRepositoryFirestore", "updateTravel")
+        performFirestoreOperation(
+            db.collection(collectionPath).document(activity.uid).set(activity), onSuccess, onFailure)
+    }
+
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt:47]: You can add comments for these function. It will clarify their purpose and usage.

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    /**
+     * This function retrieves all the events of a travel from the database. If the operation succeeds
+     * a function is applied on the list of events retrieved, otherwise a failure function is called.
+     *
+     * @param onSuccess (List<Event>) -> Unit): The function called when the retrieving of the events
+     *   went successfully.
+     * @param onFailure ((Exception) -> Unit): The function called when the retrieving of the events
+     *   fails.
+     */
+    override fun getActivity(onSuccess: (List<Activity>) -> Unit, onFailure: (Exception) -> Unit) {
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt:67]: Rename methods: Consider renaming getActivity() to getAllActivities()

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+    /**
+     * This function converts a document got from Firebase to an event. It returns null if an error
+     * occurs.
+     *
+     * @param document (DocumentSnapshot) : The document from Firebase
+     * @return (Event?) : If the conversion goes without a problem an event is return. Otherwise, null
+     *   is returned.
+     */
+    private fun documentToEvent(document: DocumentSnapshot): Activity? {
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt:131]: change documentToEvent to documentToActivity

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        return try {
+            val uid = document.id
+            val title = document.getString("title")
+            val description = document.getString("description")
+            val date = document.getTimestamp("date")
+            val documents = document.get("documentsNeeded") as? Map<String, Int>
+            val locationData = document.get("location") as? Map<*, *>
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt:138]: It would be better to do it like that for documents and locationData to provides safer type checking and conversion:
 ```kotlin
val documentsNeededData = document["documentsNeeded"] as? Map<*, *>
val documentsNeeded = documentsNeededData?.map { (key, value) -> key as String to value as Int }?.toMap()

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+    private val activities_ = MutableStateFlow<List<Activity>>(emptyList())
+    val activities: StateFlow<List<Activity>> = activities_.asStateFlow()
+
+    fun init() {
+        activityRepositoryFirebase.init { getActivities() }
+    }
+
+    fun getActivities() {
+        activityRepositoryFirebase.getActivity(onSuccess = { activities_.value = it }, onFailure = {})
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt:21]: It would be helpful to add more context to the error log in the onFailure block:

```kotlin
onFailure = { Log.e("ActivityViewModel", "Failed to add travel", it) }

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    fun deleteActivityById(activity: Activity) {
+        activityRepositoryFirebase.deleteActivityById(activity.uid, { getActivities() }, {})
+    }
+
+    fun getNewUid(): String {
+        return activityRepositoryFirebase.getNewUid()
+    }
+
+    // create factory
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt:40]: Move the Factory to the top

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+
+                }
+            ){
+                Text("Save")
+            }
+        }
+    }
+}
+
+class DateVisualTransformation: VisualTransformation {
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt:146]: It is a great idea to add a class for automatically parsing the date with slashes (/) !

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+                onValueChange = {  },
+                label = { Text("Location") },
+                placeholder = { Text("Location") }
+            )
+
+            Button(
+                enabled = location.isNotBlank() && title.isNotBlank() && description.isNotBlank()
+                        && dateText.isNotBlank(),
+                onClick = {
+                    try{
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt:103]: It would be better to consider adding more specific error handling for the date parsing and the Firebase operation failure

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+
+                        val dateFormat =
+                            SimpleDateFormat("dd/MM/yyyy", Locale.getDefault()).apply {
+                                isLenient = false // strict date format
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt:107]: it would be better to add a helper function for the parsing date

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+                            imageVector = Icons.Default.DateRange,
+                            contentDescription = null
+                        )
+                    }
+                })
+        },
+
+        bottomBar = {
+            NavigationBar {
+                listOfDestinations.forEach { NavigationBarItem(
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt:131]: I don't understand why you use a navigation bar?

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+data class BottomNavigationItem(
+    val title: String,
+    val icon: ImageVector
+)
+
+@Preview
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun TravelActivitiesScreen(
+    navigationActions: NavigationActions? = null,
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt:61]: Why did you add null as a default value for navigationActions?

**Review by Unsaved2 — CHANGES_REQUESTED** [review]
I’ve reviewed your code and didn’t find any major issues, but I do have a few suggestions:

`Activity`:
1. Add documentation for what an 'Activity' is 

`ActivityRepositoryFirebase`:

1. Add documentation for the methods updateActivity() and deleteActivityById().
2. You use 'Event' instead of 'Activity' in the documentation and for documentToEvent() 
3. Improve data handling by converting Map<*, *> to Map<String, Int> for better type safety.
4. Rename methods: Consider renaming getActivity() to getAllActivities()

`ActivityViewModel`:
1. Add selectedActivity: It would be useful to add a selectedActivity 
2. Move the Factory to the top: It's preferable to have the Factory definition at the top of the class for better organization and clarity
3. There is no documentation for this class, add some more
4. Add onfailure

`AddActivityScreen`
1. Helper function for parsing date
2. Different try catch for date and firebase operation

Good job overall!

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt (hunk 6 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt]: Yes, I agree. I did know this notation

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt (hunk 9 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt]: Thank you very much

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt (hunk 10 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt]: Can you clarify this point ? Because this operation is not related to Firebase. Before interacting with the modelView, we are tryign that the date on the text field is valid or not. For example it verifies (before calling addActivity) that the date is not 00/00/0000. so I don't see what a failure Firebase operation would do here

Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt]: Ok, no I understand. Well, we can pass a context parameter to the modelView to directly show a toast from the modelView (but displayed on the addActivity screen). And, we can separate the exception of the parsing from the failure of the model view. Good catch

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt (hunk 12 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt]: Well, it is more like a feature because I wanted to have two screens, one that displays the list of all the activities to do, and another that marks the activity of a travel on a map. And we could navigate between the two screens.

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt (hunk 13 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt:44]: We still need to discuss the navigation issue (with route and screen). So I wanted to have a functional view to work with. But, at the moment when the navigation issue is solved, I will make this parameter non-nullable.

File: app/src/main/java/com/github/se/travelpouch/model/activity/Activity.kt (hunk 1 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/activity/Activity.kt:5]: Done

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt (hunk 2 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt:41]: Done

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt (hunk 3 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt:59]: Done

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt (hunk 4 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt]: Yes, it is more explicit.

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt (hunk 5 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/activity/ActivityRepositoryFirebase.kt]: My bad. Copy pasting hazards XD.

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt (hunk 8 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt]: Done

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt (hunk 11 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt]: Yes, you are right.

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+import androidx.lifecycle.ViewModelProvider
+import com.github.se.travelpouch.model.Location
+import com.google.firebase.Firebase
+import com.google.firebase.Timestamp
+import com.google.firebase.firestore.firestore
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+
+class ActivityModelView(val activityRepositoryFirebase: ActivityRepository) : ViewModel() {
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt:16]: there is a typo : it is ActivityViewModel, no ?

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+          override fun <T : ViewModel> create(modelClass: Class<T>): T {
+            return ActivityModelView(ActivityRepositoryFirebase(Firebase.firestore)) as T
+          }
+        }
+  }
+
+  private val activities_ = MutableStateFlow<List<Activity>>(emptyList())
+  val activities: StateFlow<List<Activity>> = activities_.asStateFlow()
+
+  private val selectedActivity_ =
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt:32]: Why did you initialize selectedActivity_ with a default Activity  instead of setting it to null?

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+  /**
+   * This function adds an activity to the database
+   *
+   * @param activity (Activity) : the activity to add to the database
+   */
+  fun addActivity(activity: Activity, context: Context) {
+    activityRepositoryFirebase.addActivity(
+        activity,
+        {
+          getAllActivities()
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt:66]: for consistency and readability, wouldn't it be better to always use (onFailure = and onSuccess =) ?

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt (hunk 13 above)
Comment (Unsaved2) [reply app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt:57]: ok I see

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt (hunk 10 above)
Comment (Unsaved2) [reply app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt]: My suggestion was more about distinguishing parsing errors from activity creation errors and Firestore operations. For example, if Firestore fails, the user should receive a toast saying 'Unable to add the activity' instead of a message about an invalid date

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+
+          Button(
+              enabled =
+                  location.isNotBlank() &&
+                      title.isNotBlank() &&
+                      description.isNotBlank() &&
+                      dateText.isNotBlank(),
+              onClick = {
+                try {
+                  val finalDate = convertStringToDate(dateText, dateFormat, gregorianCalendar)
```
Comment (Unsaved2) [inline app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt:105]: I mentioned that the current approach for handling dates is fine, but it might be even better to use a function like a DatePicker to further improve the user experience.

File: app/src/androidTest/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreenTest.kt [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+    assert(result.text == "")
+    composeTestRule.onNodeWithTag("dateField").performTextInput("l09kfj89iuiui2025")
+    result = composeTestRule.onNodeWithTag("dateField").fetchSemanticsNode().config[EditableText]
+    assert(result.text == "")
+    composeTestRule.onNodeWithTag("dateField").performTextClearance()
+    composeTestRule.onNodeWithTag("dateField").performTextInput("1234")
+    composeTestRule.onNodeWithTag("dateField").performTextInput("l")
+    result = composeTestRule.onNodeWithTag("dateField").fetchSemanticsNode().config[EditableText]
+    assert(result.text == "12/34/")
+  }
```
Comment (Unsaved2) [inline app/src/androidTest/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreenTest.kt:148]: In the tests, you're checking for invalid input cases, which is great, but it would also be good to test when everything is correct and ensure that it properly calls addActivity

**Review by Unsaved2 — COMMENTED** [review]
You've implemented my previous suggestions, which is great, but I still have a few more suggestions.

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt (hunk 14 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt:16]: Yes, I inverted the two. Also, I will add some documentation

File: app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt (hunk 16 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/activity/ActivityViewModel.kt:66]: Yes, you are right

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt (hunk 17 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt:105]: Ok, we can imlpement for a later sprint. But yes, it may give a better user experience

File: app/src/androidTest/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreenTest.kt (hunk 18 above)
Comment (yassine04e) [reply app/src/androidTest/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreenTest.kt:148]: Actually, there a test, that is a bit hidden but it checks everything works when the fields are valid. It is called "doesSaveWhenFieldsAreFull" at lines 103 to 107.

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+  //              Location(0.0, 0.0, Timestamp(0, 0), "lcoation1"),
+  //              Timestamp(0, 0),
+  //              mapOf<String, Int>()),
+  //          Activity(
+  //              "2",
+  //              "title2",
+  //              "description2",
+  //              Location(0.0, 0.0, Timestamp(0, 0), "lcoation2"),
+  //              Timestamp(0, 0),
+  //              mapOf<String, Int>()))
```
Comment (RemIsMyWaifuu) [inline app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt:78]:
```suggestion
```
Here, we shouldn't leave test code once we want to merge it into main. This should stay locally.

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+      5 -> 4
+      6 -> 4
+      7 -> 5
+      8 -> 6
+      9 -> 7
+      10 -> 8
+      else -> 8
+    }
+  }
+}
```
Comment (RemIsMyWaifuu) [inline app/src/main/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreen.kt:203]: Nitpick: I think this could be easily put in another file and documented. I think this is containerisable quite easily and helps to encapsulate some complexity of visual transformers. :3

**Comment by RemIsMyWaifuu** [comment]
I think it could be nice if you had some toasts for some of the non functional buttons. For example, the navigation to the other screens or adding an activity. :D

**Review by RemIsMyWaifuu — CHANGES_REQUESTED** [review]

**Review by RemIsMyWaifuu — APPROVED** [review]
Thank you for those changes.
Keep up the good work :100:

File: app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt (hunk 12 above)
Comment (Unsaved2) [reply app/src/main/java/com/github/se/travelpouch/ui/dashboard/TravelActivityScreen.kt]: yes I understand

File: app/src/androidTest/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreenTest.kt (hunk 18 above)
Comment (Unsaved2) [reply app/src/androidTest/java/com/github/se/travelpouch/ui/dashboard/AddActivityScreenTest.kt:148]: Yes you are right

**Review by Unsaved2 — APPROVED** [review]
Thank you for taking my changes into account, everything looks good to me.

**Review by RemIsMyWaifuu — APPROVED** [review]

**Review by Unsaved2 — APPROVED** [review]
Everything is good to me 👍
